core/src/main/scala/akka/persistence/r2dbc/query/scaladsl/R2dbcReadJournal.scala

```diff
@@ -110,19 +112,52 @@ final class R2dbcReadJournal(system: ExtendedActorSystem, config: Config, cfgPat
 
   private val filteredPayloadSerId = SerializationExtension(system).findSerializerFor(FilteredPayload).identifier
 
+  // key is tuple of entity type and slice
+  private val heartbeatPersistenceIds = new ConcurrentHashMap[(String, Int), String]()
+
+  private def heartbeatPersistenceId(entityType: String, slice: Int): String = {
+    val key = entityType -> slice
+    heartbeatPersistenceIds.get(key) match {
+      case null =>
+        // no need to block other threads, it's just a cache
+        val pid = generateHeartbeatPersistenceId(entityType, slice)
+        heartbeatPersistenceIds.put(key, pid)
+        pid
+      case pid => pid
+    }
+  }
+
+  @tailrec private def generateHeartbeatPersistenceId(entityType: String, slice: Int, n: Int = 1): String = {
+    if (n < 1000000) {
+      val pid = PersistenceId.concat(entityType, s"_ht-$n")
+      if (persistenceExt.sliceForPersistenceId(pid) == slice)
+        pid
+      else
+        generateHeartbeatPersistenceId(entityType, slice, n + 1)
+    } else
+      throw new IllegalStateException(s"Couldn't find heartbeat persistenceId for [$entityType] with slice [$slice]")
+
+  }
```

Comment:
This looks like the most awkward part of the solution. Needing to find a dummy persistence id that maps to the slice.

Could we have a completely custom persistence id instead, that doesn't include the entityType? So that the persistence id still hashes to the min slice correctly, and the entityType is in the envelope already.  Then it could be generated upfront as a constant. Or do we rely on extracting the entity type from the persistence id in too many places?

Reply:
Probably not frequent enough to worry about though.

Reply:
I'm worried about that the entityType extraction (and also the slice from pid) is used in many places, since everything is scoped around the entityType, such as the query itself and sharding. Might work but high risk.

Since it's cached with the same lifecycle as R2dbcReadJournal instance (typically one per ActorSystem) it will not be very many.

Reply:
There are some tricks we could do (around how string hashcode is calculated) to avoid needing to brute force it. But that's something we could come back to later on. No problem to change what the heartbeat persistence id is, providing it still maps to the right slice.